File: packages/backend/src/storage/adminStore.ts
/**
 * Admin 账户存储模块
 * 使用 JSON 文件存储管理员账户数据
 */

import fs from 'fs'
import path from 'path'
import bcrypt from 'bcryptjs'
import { createLogger } from '../utils/logger.js'
import { getConfig } from '../config/index.js'

const logger = createLogger('AdminStore')

// 数据文件路径
const DATA_DIR = path.join(process.cwd(), 'data')
const ADMIN_FILE = path.join(DATA_DIR, 'admin.json')

// Admin 账户接口
export interface AdminUser {
  username: string
  passwordHash: string
  createdAt: number
  updatedAt: number
}

/**
 * 确保数据目录存在
 */
function ensureDataDir(): void {
  if (!fs.existsSync(DATA_DIR)) {
    fs.mkdirSync(DATA_DIR, { recursive: true })
    logger.info('Created data directory', { path: DATA_DIR })
  }
}

/**
 * 读取 Admin 数据
 */
function readAdminData(): AdminUser | null {
  try {
    if (!fs.existsSync(ADMIN_FILE)) {
      return null
    }
    const data = fs.readFileSync(ADMIN_FILE, 'utf-8')
    return JSON.parse(data) as AdminUser
  } catch (error) {
    logger.error('Failed to read admin data', { error: (error as Error).message })
    return null
  }
}

/**
 * 写入 Admin 数据
 */
function writeAdminData(admin: AdminUser): void {
  try {
    ensureDataDir()
    fs.writeFileSync(ADMIN_FILE, JSON.stringify(admin, null, 2), 'utf-8')
    logger.info('Admin data saved')
  } catch (error) {
    logger.error('Failed to write admin data', { error: (error as Error).message })
    throw error
  }
}

/**
 * 初始化 Admin 账户
 * 如果不存在则从环境变量创建默认账户
 */
export async function initAdmin(): Promise<void> {
  const existing = readAdminData()

  if (existing) {
    logger.info('Admin account already exists', { username: existing.username })
    return
  }

  const config = getConfig()
  const { username, password } = config.admin

  const salt = await bcrypt.genSalt(10)
  const passwordHash = await bcrypt.hash(password, salt)

  const admin: AdminUser = {
    username,
    passwordHash,
    createdAt: Date.now(),
    updatedAt: Date.now()
  }

  writeAdminData(admin)
  logger.info('Admin account initialized', { username })
}

/**
 * 获取 Admin 账户
 */
export function getAdmin(): AdminUser | null {
  return readAdminData()
}

/**
 * 验证密码
 */
export async function validatePassword(username: string, password: string): Promise<boolean> {
  const admin = readAdminData()

  if (!admin) {
    logger.warn('Admin account not found')
    return false
  }

  if (admin.username !== username) {
    logger.warn('Invalid username', { provided: username })
    return false
  }

  const isValid = await bcrypt.compare(password, admin.passwordHash)

  if (!isValid) {
    logger.warn('Invalid password for user', { username })
  }

  return isValid
}

/**
 * 更新密码
 */
export async function updatePassword(oldPassword: string, newPassword: string): Promise<boolean> {
  const admin = readAdminData()

  if (!admin) {
    logger.error('Admin account not found')
    return false
  }

  // 验证旧密码
  const isValid = await bcrypt.compare(oldPassword, admin.passwordHash)
  if (!isValid) {
    logger.warn('Old password verification failed')
    return false
  }

  // 生成新密码哈希
  const salt = await bcrypt.genSalt(10)
  const passwordHash = await bcrypt.hash(newPassword, salt)

  // 更新账户
  admin.passwordHash = passwordHash
  admin.updatedAt = Date.now()

  writeAdminData(admin)
  logger.info('Admin password updated')

  return true
}

File: packages/gateway/storage/admin_store.go
package storage

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sync"
	"time"

	"github.com/rs/zerolog/log"
	"golang.org/x/crypto/bcrypt"
)

// AdminUser represents the persisted admin account.
type AdminUser struct {
	Username     string `json:"username"`
	PasswordHash string `json:"passwordHash"`
	CreatedAt    int64  `json:"createdAt"`
	UpdatedAt    int64  `json:"updatedAt"`
}

var (
	adminMu   sync.RWMutex
	adminFile string
)

func init() {
	adminFile = filepath.Join("data", "admin.json")
}

// SetAdminFilePath allows overriding the admin file path (e.g. for testing).
func SetAdminFilePath(path string) {
	adminMu.Lock()
	defer adminMu.Unlock()
	adminFile = path
}

// ensureDataDir creates the data directory if it does not exist.
func ensureDataDir() error {
	dir := filepath.Dir(adminFile)
	if err := os.MkdirAll(dir, 0755); err != nil {
		return fmt.Errorf("create data dir: %w", err)
	}
	return nil
}

// readAdminData reads the admin user from the JSON file.
func readAdminData() (*AdminUser, error) {
	adminMu.RLock()
	defer adminMu.RUnlock()

	data, err := os.ReadFile(adminFile)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, fmt.Errorf("read admin file: %w", err)
	}

	var admin AdminUser
	if err := json.Unmarshal(data, &admin); err != nil {
		return nil, fmt.Errorf("unmarshal admin data: %w", err)
	}
	return &admin, nil
}

// writeAdminData writes the admin user to the JSON file.
func writeAdminData(admin *AdminUser) error {
	adminMu.Lock()
	defer adminMu.Unlock()

	if err := ensureDataDir(); err != nil {
		return err
	}

	data, err := json.MarshalIndent(admin, "", "  ")
	if err != nil {
		return fmt.Errorf("marshal admin data: %w", err)
	}

	if err := os.WriteFile(adminFile, data, 0644); err != nil {
		return fmt.Errorf("write admin file: %w", err)
	}

	log.Info().Msg("Admin data saved")
	return nil
}

// InitAdmin creates the admin account from the given credentials if it does
// not already exist. This matches the TypeScript backend behavior.
func InitAdmin(username, password string) error {
	existing, err := readAdminData()
	if err != nil {
		return err
	}
	if existing != nil {
		log.Info().Str("username", existing.Username).Msg("Admin account already exists")
		return nil
	}

	hash, err := bcrypt.GenerateFromPassword([]byte(password), bcrypt.DefaultCost)
	if err != nil {
		return fmt.Errorf("hash password: %w", err)
	}

	now := time.Now().UnixMilli()
	admin := &AdminUser{
		Username:     username,
		PasswordHash: string(hash),
		CreatedAt:    now,
		UpdatedAt:    now,
	}

	if err := writeAdminData(admin); err != nil {
		return err
	}

	log.Info().Str("username", username).Msg("Admin account initialized")
	return nil
}

// ValidatePassword checks the provided credentials against the stored admin.
func ValidatePassword(username, password string) (bool, error) {
	admin, err := readAdminData()
	if err != nil {
		return false, err
	}
	if admin == nil {
		log.Warn().Msg("Admin account not found")
		return false, nil
	}
	if admin.Username != username {
		log.Warn().Str("provided", username).Msg("Invalid username")
		return false, nil
	}

	if err := bcrypt.CompareHashAndPassword([]byte(admin.PasswordHash), []byte(password)); err != nil {
		log.Warn().Str("username", username).Msg("Invalid password")
		return false, nil
	}
	return true, nil
}

// UpdatePassword changes the admin password after verifying the old one.
func UpdatePassword(oldPassword, newPassword string) error {
	admin, err := readAdminData()
	if err != nil {
		return err
	}
	if admin == nil {
		return fmt.Errorf("admin account not found")
	}

	if err := bcrypt.CompareHashAndPassword([]byte(admin.PasswordHash), []byte(oldPassword)); err != nil {
		return fmt.Errorf("old password verification failed")
	}

	hash, err := bcrypt.GenerateFromPassword([]byte(newPassword), bcrypt.DefaultCost)
	if err != nil {
		return fmt.Errorf("hash new password: %w", err)
	}

	admin.PasswordHash = string(hash)
	admin.UpdatedAt = time.Now().UnixMilli()

	if err := writeAdminData(admin); err != nil {
		return err
	}

	log.Info().Msg("Admin password updated")
	return nil
}

// GetAdmin returns the current admin user, or nil if none exists.
func GetAdmin() (*AdminUser, error) {
	return readAdminData()
}
